docs/source/torch.compiler_dynamic_shapes.md

```diff
@@ -1,129 +1,474 @@
-- Sometimes it can be cumbersome to find the right inputs to mark as dynamic. If
-  you're willing to take a performance hit for the first batch, one other affordable
-  option we have are the eager_then_compile stances which derive dynamism for you.
-  See [torch.compiler.set_stance](https://docs.pytorch.org/docs/stable/generated/torch.compiler.set_stance.html) for more details.
+This setting forces all sizes and integers to be dynamic, increasing the
+chance of encountering dynamic shape bugs. Setting this option is not recommended due to it  being error prone.
+It would make every input size dynamic which may result it performance regressions and ultimately increase compilation time.
 
-## The Guard Model
+### Profile-Guided Optimization (PGO)
 
-When considering how to add support for dynamic shapes to TorchDynamo and TorchInductor, we made a major design decision: in order to reuse decompositions and other preexisting code written in Python/C++ targeting the PyTorch API, we must be able to trace through dynamic shapes. Unlike a fully symbolic system which might capture both branches of a conditional, we always pick one branch and specialize our trace under the assumption that we only use this trace when we would have made the same choice for that branch in the future. To do this, we maintain a "hint" for every symbolic size saying what its concrete value is at compile time (as TorchDynamo is a just-in-time compiler, it always knows what the actual input sizes are.) When we perform a condition on a tensor, we simply consult the hint to find out which branch to take.
+Profile-Guided Optimization (PGO) extends automatic dynamic across attempts, learning from previous runs to avoid initial static compilations. This means that things marked as dynamic in attempt 1 will remain dynamic in attempt 2 from the first compilation. If attempt 2 encounters different sizes for the same input, they will be marked as dynamic.
 
-This greatly simplifies the symbolic shape formulas we produce, but means we have a much more involved system for managing guards. Consider, for example, the following program:
+For example, for the program discussed earlier:
 
 ```python
-def f(x, y):
-    z = torch.cat([x, y])
-    if z.size(0) > 2:
-        return z.mul(2)
-    else:
-        return z.add(2)
+def f(x):
+    return x * x.size()[0]
+```
+
+In attempt 0, when we first encounter `f(x)` with `x=10`, we make it
+static since it's the initial observation. The second time we encounter `f(x)`
+with `x=20`, we have observed two different
+values for `x`, so we make it dynamic through automatic dynamic behavior.
+
+In attempt 1, we repeat the process above unless Profile-Guided Optimization (PGO) is enabled.
+With PGO, we already know from attempt 0 that `f(x)` should be dynamic, so it is marked as
+such the first time we encounter it.
+
+(identifying-dynamic-elements-marked-by-pgo)=
+#### Identifying Dynamic Elements Marked by PGO
```

Comment:
I think this can probably be omitted/extracted too. At the very least if we are going to go down this route, we should close the loop and show what the "last step" is, presumably either via mark_dynamic or the allowlist.

Reply:
removed